crates/rust-analyzer/src/handlers/request.rs

```diff
@@ -1995,14 +1996,52 @@ fn run_rustfmt(
             cmd
         }
         RustfmtConfig::CustomCommand { command, args } => {
-            let mut cmd = process::Command::new(command);
+            let cmd = PathBuf::from(&command);
+            let mut components = cmd.components();
+
+            // to support rustc's suggested, default configuration
+            let mut cmd = match components.next() {
+                Some(std::path::Component::CurDir) => {
+                    let rest = components.as_path();
+
+                    let roots = snap
+                        .workspaces
+                        .iter()
+                        .flat_map(|ws| ws.workspace_definition_path())
+                        .collect::<Vec<&AbsPath>>();
+
+                    let abs: Option<AbsPathBuf> = roots.into_iter().find_map(|base| {
+                        let abs = base.join(rest);
+                        std::fs::metadata(&abs).ok().map(|_| abs)
+                    });
```

Comment:
I'm not a fan of this approach, but I can't find a better way to determine what workspace a `TextDocument` is a part of.

Reply:
Take a look at `CargoTargetSpec::for_file`, that function has to solve the same problem of relating a file to its workspace(s)